src/Chainweb/BlockHeaderDB/PruneForks.hs

```diff
@@ -41,180 +55,199 @@ import System.LogLevel
+-- of blocks.
 --
 -- This function guarantees that the predecessors of all remaining nodes also
 -- remain in the database and that there are are no tangling references.
 --
 -- This function doesn't guarantee to delete all blocks on forks. A small number
 -- of fork blocks may not get deleted.
 --
--- The function takes a callback that is invoked on each deleted block header.
--- The callback takes a parameter that indicates whether the block payload hash
--- is shared with any non-deleted block header. There is a small rate of false
--- positives of block payload hashes that are marked in use.
---
--- This doesn't update the the cut db or the payload db.
+-- The function takes a callback that are invoked on each block header in the
+-- database starting from the given depth. The callback takes a parameter that
+-- indicates whether the related block is pruned or not.
 --
 pruneForks
-    :: Logger logger
-    => logger
+    :: LogFunctionText
     -> BlockHeaderDb
     -> Natural
-        -- ^ The depth at which the roots are collected for marking block
-        -- headers that are kept. Any fork that was active that the given depth
-        -- is kept.
-        --
-        -- The depth is computed based on the entry of maximum rank. This entry
-        -- is not necessarily included in the overall consensus of the chainweb.
-        -- In particular the the higest ranking entry is not necessarily the
-        -- entry of largest POW weight.
-        --
-        -- Usually this number would be defined in terms of the chainweb
-        -- diameter and/or the difficulty adjustment window.
-    -> (BlockHeader -> Bool -> IO ())
+        -- ^ The depth at which deletion starts. Note, that the max rank isn't
+        -- necessarly included in the current best cut. So one, should choose a
+        -- depth for which one is confident that all forks are resolved.
+
+    -> (Bool -> BlockHeader -> IO ())
         -- ^ Deletion call back. This hook is called /after/ the entry is
         -- deleted from the database. It's main purpose is to delete any
         -- resources that were related to the deleted header and that are not
-        -- needed any more. The Boolean argument indicates whether the payload
-        -- of the block is shared with any block header that isn't marked for
-        -- deletion.
+        -- needed any more. The first parameter indicates whether the block
+        -- header got deleted from the chain database.
+
     -> IO Int
-pruneForks logger = pruneForks_ logg
+pruneForks logg cdb depth callback = do
+    hdr <- maxEntry cdb
+    if
+        | int (_blockHeight hdr) <= depth -> do
+            logg Info
+                $ "Skipping database prunning because the requested depth "
+                <> sshow depth <> " is not larger than the maximum block height "
+                <> sshow (_blockHeight hdr)
+            return 0
+        | int (_blockHeight hdr) <= int genHeight + depth -> do
```

Comment:
Can you document how this case gets tested (I couldn't find anything in the manual tests)?

Reply:
This gets triggered for instance on the mainnet database for the new chains. I haven seen that showing up on the mainnet-simulation